### PR TITLE
[*] WS : Optionally send in-transit email when updating tracking number

### PR DESCRIPTION
**Pull request by gerdus**
| Questions | Answers |
| --- | --- |
| Branch? | develop |
| Description? | Optionally send in-transit email when updating tracking number. Moved the email sending to OrderCarrier from admin order controller in order to accomplish this. |
| Type? | improvement |
| Category? | WS |
| BC breaks? | no |
| Deprecations? | no |
| Fixed ticket? | n/a |
| How to test? | add sendemail=1 to URL e.g. myshop.com/api/order_carriers/7?sendemail=1 and do a PUT (update) |

### CODE REVIEW

**Comment by gerdus** [comment]
sorry about the many commits in my branch history. I am a git noob. If those commits can't be squashed when merging the pull request I'll redo the changes on a new branch and submit the pull request again.

**Comment by firstred** [comment]
Don't worry. Thank you for your contribution!

You can easily squash the commits with an interactive rebase yourself:
I assume that you have 6 commits here: 
```shell
git rebase -i HEAD~5
```
Make sure that the first one is set to `pick` and the others `squash`
Don't forget to (force) push
```shell
git push -f
```

**Comment by gerdus** [comment]
Thanks @firstred you set me on the right path.

File: classes/order/OrderCarrier.php [hunk 1]
```diff
@@ -47,7 +47,7 @@ class OrderCarrierCore extends ObjectModel
     /** @var float */
     public $shipping_cost_tax_incl;
 
-    /** @var int */
+    /** @var string */
```
Comment (gerdus) [inline classes/order/OrderCarrier.php]: fix type comment

File: classes/order/OrderCarrier.php [hunk 2]
```diff
@@ -72,9 +72,66 @@ class OrderCarrierCore extends ObjectModel
...
+        }
+        if (!Validate::isLoadedObject($carrier)) {
+            throw new PrestaShopException('Can\'t load Carrier object');
+        }
+        $templateVars = array(
+            '{followup}' => str_replace('@', $this->tracking_number, $carrier->url),
+            '{firstname}' => $customer->firstname,
+            '{lastname}' => $customer->lastname,
+            '{id_order}' => $order->id,
+            '{shipping_number}' => $this->tracking_number,
```
Comment (gerdus) [inline classes/order/OrderCarrier.php]: kept name of the template var for tracking number "shipping_number" to preserve template compatability

**Comment by gerdus** [comment]
rebased the pull request

**Comment by xBorderie** [comment]
Thank you for rebasing! Currently the team is mostly working on urgent/blocking issue that have surfaced since the 1.7.0.0. Your PR will be reviewed soon!

File: classes/order/OrderCarrier.php [hunk 3]
```diff
@@ -72,9 +72,84 @@ class OrderCarrierCore extends ObjectModel
...
+        if (!Validate::isLoadedObject($carrier)) {
+            throw new PrestaShopException('Can\'t load Carrier object');
+        }
+        $orderLanguage = new Language((int) $order->id_lang);
+        $templateVars = array(
+            '{followup}' => str_replace('@', $order->shipping_number, $carrier->url),
+            '{firstname}' => $customer->firstname,
+            '{lastname}' => $customer->lastname,
+            '{id_order}' => $order->id,
+            '{shipping_number}' => $this->tracking_number,
```
Comment (aleeks) [inline classes/order/OrderCarrier.php]: Why $this & not $order here ?

**Comment by aleeks** [comment]
Hello  @gerdus 
To me, it looks good, just the comment ? why $this and not $order ? for the tracking number ?

**Comment by aleeks** [comment]
Hello @gerdus 
I cherry-pick your commit here: https://github.com/PrestaShop/PrestaShop/pull/7508 with an other PR.
Regards

File: classes/order/OrderCarrier.php (hunk 3 above)
Comment (gerdus) [inline classes/order/OrderCarrier.php]: because shipping_number on Order is deprecated:
`@deprecated 1.5.0.4
@see OrderCarrier->tracking_number
`

**Pull request closed without merging**